### PR TITLE
Python: Add PyRelation.fetchnumpy()

### PR DESCRIPTION
**Pull request by wisp3rwind**
This is just the new `PyRelation.fetchnumpy()` method from https://github.com/duckdb/duckdb/pull/4099. Tests (hopefully, I found it hard to see where exactly to put the tests) as suggested in

> For sure your fetch numpy code, try to hit all the lines of code with the tests (so also an invalid query), and maybe do empty result, < vector_size, > vector_size (vector_size is 1024 elements)

_Originally posted by @pdet in https://github.com/duckdb/duckdb/issues/4099#issuecomment-1181499189_

The PR is relative to #4427; and should be easy to review commit-by-commit.

### CODE REVIEW

File: tools/pythonpkg/tests/fast/test_all_types.py [hunk 1]
```diff
@@ -84,6 +86,253 @@ def test_fetchall(self, duckdb_cursor):
...
+        # The following types lead to errors:
+        # Conversion Error: Date out of range in timestamp conversion
+        # - 'timestamp_array'
+        # - 'timestamptz_array'
+        # Conversion Error: Could not convert Timestamp(US) to Timestamp(NS)
+        # - 'timestamp'
+        # - 'timestamp_s'
+        # - 'timestamp_ms'
+        # - 'timestamp_tz'
+        # SystemError: <built-in function __import__> returned a result with an exception set
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/test_all_types.py:313]: This definitely sounds like something we should investigate further

File: tools/pythonpkg/tests/fast/test_all_types.py [hunk 2]
```diff
@@ -84,6 +86,253 @@ def test_fetchall(self, duckdb_cursor):
...
+
+        rel = conn.table_function("test_all_types")
+        for cur_type in all_types:
+            if cur_type not in correct_answer_map:
+                continue
+
+            result = rel.project(cur_type).fetchnumpy()
+            result = result[cur_type]
+            correct_answer = correct_answer_map[cur_type]
+            print(cur_type)
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/test_all_types.py:324]: Leftover print?

File: tools/pythonpkg/tests/fast/test_relation.py [hunk 3]
```diff
@@ -200,3 +200,33 @@ def test_explain(self, duckdb_cursor):
...
+            # For some reason, this return a masked array. Shouldn't it be
+            # known that there can't be NULLs?
+            if isinstance(res, np.ma.MaskedArray):
+                assert res.count() == size
+                res = res.compressed()
+            else:
+                assert len(res["a"]) == size
+            assert np.all(res["a"] == np.arange(start, start + size))
+
+        with pytest.raises(Exception):
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/test_relation.py:229]: Please add the exception message as the match parameter here

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/tests/fast/test_all_types.py (hunk 1 above)
Comment (wisp3rwind) [reply tools/pythonpkg/tests/fast/test_all_types.py:313]: It's unrelated to the PR:

```
Python 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21)
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import duckdb
>>> print(duckdb.__version__)
0.4.0
>>> con = duckdb.connect()
>>> t = con.table_function("test_all_types")
>>> t.project("date_array").execute().fetchnumpy()
SystemError: /home/conda/feedstock_root/build_artifacts/python-split_1653669925074/work/Objects/listobject.c:318: bad argument to internal function

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: <built-in function __import__> returned a result with an error set
>>>
```

Might be appending a `NULL` to a Python list? Cf. https://github.com/python/cpython/blob/d348afa15d5a997e7a8e51c0f789f41cb15cc651/Objects/listobject.c#L318

File: tools/pythonpkg/tests/fast/test_all_types.py (hunk 2 above)
Comment (wisp3rwind) [reply tools/pythonpkg/tests/fast/test_all_types.py]: Yes, fixed.

File: tools/pythonpkg/tests/fast/test_relation.py (hunk 3 above)
Comment (wisp3rwind) [reply tools/pythonpkg/tests/fast/test_relation.py]: Done

File: tools/pythonpkg/tests/fast/test_all_types.py (hunk 1 above)
Comment (Mause) [reply tools/pythonpkg/tests/fast/test_all_types.py:313]: Yeah I didn't think it was related to this pr, just that we should investigate

**Review by Mause — APPROVED** [review]

**Comment by hannes** [comment]
Can you resolve the conflict please? Otherwise good to go.

File: tools/pythonpkg/tests/fast/test_relation.py [hunk 4]
```diff
@@ -200,3 +200,33 @@ def test_explain(self, duckdb_cursor):
...
+            # For some reason, this return a masked array. Shouldn't it be
+            # known that there can't be NULLs?
+            if isinstance(res, np.ma.MaskedArray):
+                assert res.count() == size
+                res = res.compressed()
+            else:
+                assert len(res["a"]) == size
+            assert np.all(res["a"] == np.arange(start, start + size))
+
+        with pytest.raises(Exception, match="Conversion Error.*out of range.*"):
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/test_relation.py:229]:
```suggestion
        with pytest.raises(duckdb.ConversionException, match="Conversion Error.*out of range.*"):
```

**Review by Mause — CHANGES_REQUESTED** [review]

**Review by Mause — APPROVED** [review]

**Comment by hannes** [comment]
Looks good thanks

**Pull request merged**